internal/pkg/service/common/servicectx/servicectx_test.go

```diff
@@ -15,71 +16,83 @@ import (
 func TestProcess_Add(t *testing.T) {
 	t.Parallel()
 
-	ctx, cancel := context.WithCancel(context.Background())
 	logger := log.NewDebugLogger()
-	proc, err := New(ctx, cancel, WithLogger(logger), WithUniqueID("<id>"))
+	logger.ConnectTo(os.Stdout)
+	proc, err := New(WithLogger(logger), WithUniqueID("<id>"))
 	assert.NoError(t, err)
 
-	op1 := &sync.WaitGroup{}
-	op1.Add(1)
-	op2 := &sync.WaitGroup{}
-	op2.Add(1)
-	op3 := &sync.WaitGroup{}
-	op3.Add(1)
-
-	// Do some work, operations run in parallel
-	proc.Add(func(ctx context.Context, shutdown ShutdownFn) {
-		<-ctx.Done()
-		logger.InfoCtx(ctx, "end1")
-		op1.Done()
+	// OpCtx simulates long running operations
+	opCtx, opCancel := context.WithCancel(context.Background())
+
+	// There are some parallel operations
+	opWg := &sync.WaitGroup{}
+	opWg.Add(3)
+	proc.Add(func(shutdown ShutdownFn) {
+		<-opCtx.Done()
+		logger.InfoCtx(opCtx, "end") // STEP 4, see <-opCtx.Done()
+		opWg.Done()
 	})
-	proc.Add(func(ctx context.Context, shutdown ShutdownFn) {
-		<-ctx.Done()
```

Comment:
The `Add` method originally provided the "root" context (which ends when the shutdown was requested).

But in practice we didn't use it like that:

For example, there are two sub-contexts, which must be closed in a specified order:
https://github.com/keboola/keboola-as-code/blob/7404898578f8b28f17f30a3bf08ac0cf46984b90/internal/pkg/service/common/distribution/node.go#L64-L75